lib/compilers/spirv-tools.ts

```diff
@@ -0,0 +1,176 @@
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import path from 'path';
+
+import type {ExecutionOptionsWithEnv} from '../../types/compilation/compilation.interfaces.js';
+import type {PreliminaryCompilerInfo} from '../../types/compiler.interfaces.js';
+import type {ParseFiltersAndOutputOptions} from '../../types/features/filters.interfaces.js';
+import {BaseCompiler} from '../base-compiler.js';
+import {CompilationEnvironment} from '../compilation-env.js';
+import {logger} from '../logger.js';
+import {SPIRVAsmParser} from '../parsers/asm-parser-spirv.js';
+import * as utils from '../utils.js';
+
+// If you want to output SPIR-V, most likely you want SPIRVAsmParser
+//
+// SPIR-V is an IR that targets heterogeneous (like a GPU) and has a "compute" and "graphics" mode.
+// When used for graphics, it inself is not enough info to "compile" down to the GPU's assembly because normally there
+// is other graphics // pipeline information that is required (ex. is the depth test enable/disabled)
+//
+// SPIR-V has a lot of tooling around it to optimize, validate, fuzz, etc. This compiler is only used for tooling.
+export class SPIRVToolsCompiler extends BaseCompiler {
+    protected assemblerPath: string;
+    protected disassemblerPath: string;
+
+    static get key() {
+        return 'spirv-tools';
+    }
+
+    constructor(compilerInfo: PreliminaryCompilerInfo, env: CompilationEnvironment) {
+        super(compilerInfo, env);
+
+        this.asm = new SPIRVAsmParser(this.compilerProps);
+
+        // spirv-as
+        this.assemblerPath = this.compilerProps<string>('assemblerPath');
+        // spirv-dis
+        this.disassemblerPath = this.compilerProps<string>('disassemblerPath');
+    }
+
+    override optionsForFilter(filters: ParseFiltersAndOutputOptions, outputFilename: string) {
+        const sourceDir = path.dirname(outputFilename);
+        const spvBinFilename = this.getPrimaryOutputFilename(sourceDir, this.outputFilebase);
+        return ['-o', spvBinFilename];
+    }
+
+    getPrimaryOutputFilename(dirPath: string, outputFilebase: string) {
+        return path.join(dirPath, `${outputFilebase}.spv`);
+    }
+
+    override getOutputFilename(dirPath: string, outputFilebase: string) {
+        return path.join(dirPath, `${outputFilebase}.spvasm`);
+    }
+
+    // Some tools (ex. spirv-opt) needs this as a single argument, so merge with "="
+    mergeSpirvTargetEnv(options: string[]) {
+        const index = options.indexOf('--target-env');
+        if (index !== -1) {
+            options.splice(index, 2, `--target-env=${options[index + 1]}`);
+        }
+        return options;
+    }
+
+    // Some tools (ex. spirv-val) needs this as a two argument, so unmerge with "="
```

Comment:
Ditto. Also what a shame! 😁

Reply:
ya... been a low hanging task to fix and support both, one day ... lol